plugins/builds/index.js

```diff
@@ -812,115 +854,165 @@ exports.register = (server, options, next) => {
              * 3. ([~D,B,C]-> sd@123:A) currentJob=D, nextJob=sd@123:A, joinList(A)=[sd@111:B,sd@111:C]
              *    joinList doesn't include sd@111:D, so start A
              */
-            // Note: current job can be "external" in nextJob's perspective
-            const currentJobNotInJoinList = !joinListNames.includes(currentJobName) &&
-                !joinListNames.includes(`sd@${pipelineId}:${currentJobName}`);
-
-            // Handle no join case
             if (joinListNames.length === 0 || currentJobNotInJoinList) {
+                // Next build is internal
                 if (!isExternal) {
+                    const internalBuildConfig = {
+                        jobFactory,
+                        buildFactory,
+                        eventFactory,
+                        pipelineId,
+                        jobName: nextJobName,
+                        username,
+                        scmContext,
+                        build, // this is the parentBuild for the next build
+                        baseBranch: event.baseBranch || null,
+                        parentBuilds
+                    };
+
                     return createInternalBuild(internalBuildConfig);
                 }
 
-                return createExternalBuild(externalBuildConfig).then(async (externalEvent) => {
-                    const externalJobId = workflowGraph.nodes
-                        .find(n => n.name === nextJobName).id;
-                    const externalBuildId = externalEvent.builds
-                        .find(b => b.jobId === externalJobId).id;
-                    const externalBuild = await buildFactory.get({ id: externalBuildId });
-                    const masterParentBuilds = {};
-
-                    masterParentBuilds[externalEvent.pipelineId] = {
-                        eventId: externalEvent.id,
-                        jobs: { [externalJobName]: externalBuildId }
-                    };
-                    const nextJobToTrigger = workflowParser.getNextJobs(workflowGraph,
-                        { trigger: nextJobName, chainPR: pipeline.chainPR });
+                console.log('no join, not internal');
+
+                /* GET OR CREATE NEXT BUILD, UPDATE WITH PARENT BUILDS INFO, AND
+                 * DECIDE IF NEED TO START
+                 * If next job is an external join job (if parentBuilds pipelineId
+                 * matches next external job pipelineId), get build and start it
+                 * if previous required builds are done successfully.
+                 * Otherwise, create internal build for matching pipeline
+                 */
+                if (build.parentBuilds && build.parentBuilds[externalPipelineId]) {
+                    console.log(`parentBuilds matches for build ${build.id}`);
+                    const externalEventId = build.parentBuilds[externalPipelineId].eventId;
+                    const externalEvent = await eventFactory.get(externalEventId);
```

Comment:
we should add a note to refactor these later, all these `await` are multiple DB calls.

Reply:
added